### DISCUSSION TITLE
Generic type constraint: notabstract

### DISCUSSION THREAD

**Discussion by ActivistInvestor** [dotnet/runtime#102581] [Ideas]
As per the title, I'm often confronted with the problem of not being able to constrain a generic parameter to a type that is not abstract. the new() constraint (which without being able to specify parameters, is about as useful as tits on a bull) doesn't work because I don't require a public parameter-less constructor. Passing an abstract type as a generic parameter when the type doesn't allow it shouldn't require a runtime check/error.

```
class Foo<T> where T: SomeBaseType, notabstract
{
}
```

That's basically it it in a nutshell.

**Comment by elgonzo** [comment]
I don't understand.

Unless you want to create an instance of `T` in your generic class, it wouldn't matter whether `T` is abstract or not.
But if you want to create an instance of `T` in your generic class, how would the generic class know the signature of the constructor(s) of `T`?

**Reply by ActivistInvestor** [reply]
>> Unless you want to create an instance of T in your generic class, it wouldn't matter whether T is abstract or not.

I'm not sure what gives you that Idea.  

```
public abstract class Base
{
   public string SomeProperty { get; set; }
}

public static class Provider<T> where T : Base
{
   static T instance = default(T);
   static Func<T> factory;

   public static void Initialize(Func<T> factory)
   {
      Provider<T>.factory = factory;
   }

   public static T Instance 
   {
      get 
      {
         if(instance == null)
         {
            instance = factory();
            /// Provider is responsible for initializing this:
            instance.SomeProperty = "some value";
         }
         return instance;
      } 
   }
}


```

**Comment by CyrusNajmabadi** [comment]
This can be done by markign the type parameter with an attribute, and having an analyzer check that attribute is fulfilled at the generic instantiation point.

**Comment by ActivistInvestor** [comment]
I can't Mark the type with an attribute. In the example I showed the type Base is a type that I don't own, and can't access.
